src/token_map_impl.hpp

```diff
@@ -0,0 +1,777 @@
+struct hash<cass::Host::Ptr> {
+  std::size_t operator()(const cass::Host::Ptr& host) const {
+    if (!host) return 0;
+    return hash(host->address());
+  }
+  std::hash<cass::Address> hash;
+};
+
+template<>
+struct equal_to<cass::Host::Ptr> {
+  bool operator()(const cass::Host::Ptr& lhs, const cass::Host::Ptr& rhs) const {
+    if (lhs == rhs) {
+      return true;
+    }
+    if (!lhs || !rhs) {
+      return false;
+    }
+    return lhs->address() == rhs->address();
+  }
+};
+
+} // namespace std
+
+namespace cass {
+
+class IdGenerator {
+public:
+  typedef sparsehash::dense_hash_map<std::string, uint32_t> IdMap;
+
+  static const uint32_t EMPTY_KEY;
+  static const uint32_t DELETED_KEY;
+
+  IdGenerator() {
+    ids_.set_empty_key(std::string());
+  }
+
+  uint32_t get(const std::string& key) {
+    if (key.empty()) {
+      return 0;
+    }
+
+    IdMap::const_iterator i = ids_.find(key);
+    if (i != ids_.end()) {
+      return i->second;
+    }
+
+    // This will never generate a 0 identifier. So 0 can be used as
+    // inalid or empty.
+    uint32_t id = ids_.size() + 1;
+    ids_[key] = id;
+    return id;
+  }
+
+private:
+  IdMap ids_;
+};
+
+struct Murmur3Partitioner {
+  typedef int64_t Token;
+
+  static Token from_string(const StringRef& str);
+  static Token hash(const StringRef& str);
+  static StringRef name() { return "Murmur3Partitioner"; }
+};
+
+struct RandomPartitioner {
+  struct Token {
+    uint64_t hi;
+    uint64_t lo;
+
+    bool operator<(const Token& other) const {
+      return hi == other.hi ? lo < other.lo : hi < other.hi;
+    }
+
+    bool operator==(const Token& other) const {
+      return hi == other.hi && lo == other.lo;
+    }
+  };
+
+  static Token from_string(const StringRef& str);
+  static Token hash(const StringRef& str);
+  static StringRef name() { return "RandomPartitioner"; }
+};
+
+class ByteOrderedPartitioner {
+public:
+  typedef std::vector<uint8_t> Token;
+
+  static Token from_string(const StringRef& str);
+  static Token hash(const StringRef& str);
+  static StringRef name() { return "ByteOrderedPartitioner"; }
+};
+
+class HostSet : public sparsehash::dense_hash_set<Host::Ptr> {
+public:
+  HostSet() {
+    set_empty_key(Host::Ptr(new Host(Address::EMPTY_KEY, false)));
+    set_deleted_key(Host::Ptr(new Host(Address::DELETED_KEY, false)));
+  }
+};
+
+class RackSet : public sparsehash::dense_hash_set<uint32_t> {
+public:
+  RackSet() {
+    set_empty_key(IdGenerator::EMPTY_KEY);
+    set_deleted_key(IdGenerator::DELETED_KEY);
+  }
+};
+
+struct Datacenter {
+  size_t num_nodes;
+  RackSet racks;
+};
+
+class DatacenterMap : public sparsehash::dense_hash_map<uint32_t, Datacenter> {
+public:
+  DatacenterMap() {
+    set_empty_key(IdGenerator::EMPTY_KEY);
+    set_deleted_key(IdGenerator::DELETED_KEY);
+  }
+};
+
+struct ReplicationFactor {
+  size_t count;
+  std::string name; // Used for logging the datacenter name
+  bool operator==(const ReplicationFactor& other) const {
```

Comment:
Should this consider DC name? Wondering omitting it would allow strategies with `{'dc1': '2', 'dc2': '2'} == {'dc1': '2', 'dc3': '2'}` to compare `true`.